adapters/pubmatic/pubmatic.go

```diff
@@ -448,6 +441,118 @@ func getNativeAdm(adm string) (string, error) {
+	if source != nil {
+		dataMap := make(map[string]interface{})
+		err := json.Unmarshal(source, &dataMap)
+		if err == nil {
+			return dataMap
+		}
+	}
+	return nil
+}
+
+//populateFirstPartyDataImpAttributes will parse imp.ext.data and populate imp extMap
+func populateFirstPartyDataImpAttributes(data json.RawMessage, extMap map[string]interface{}) {
+
+	dataMap := getMapFromJSON(data)
+
+	if dataMap == nil {
+		return
+	}
+
+	populateAdUnitKey(data, dataMap, extMap)
+	populateDctrKey(dataMap, extMap)
+}
+
+//populateAdUnitKey parses data object to read and populate DFP adunit key
+func populateAdUnitKey(data json.RawMessage, dataMap, extMap map[string]interface{}) {
+
+	if name, err := jsonparser.GetString(data, "adserver", "name"); err == nil && name == AdServerGAM {
+		if adslot, err := jsonparser.GetString(data, "adserver", "adslot"); err == nil && adslot != "" {
+			extMap[ImpExtAdUnitKey] = adslot
+		}
+	}
+
+	//imp.ext.dfp_ad_unit_code is not set, then check pbadslot in imp.ext.data
+	if extMap[ImpExtAdUnitKey] == nil && dataMap[PBAdslotKey] != nil {
+		extMap[ImpExtAdUnitKey] = dataMap[PBAdslotKey].(string)
+	}
+}
+
+//populateDctrKey reads key-val pairs from imp.ext.data and add it in imp.ext.key_val
+func populateDctrKey(dataMap, extMap map[string]interface{}) {
+	var dctr strings.Builder
+
+	//append dctr key if already present in extMap
+	if extMap[dctrKeyName] != nil {
+		dctr.WriteString(extMap[dctrKeyName].(string))
+	}
+
+	for key, val := range dataMap {
+
+		//ignore 'pbaslot' and 'adserver' key as they are not targeting keys
+		if key == PBAdslotKey || key == AdServerKey {
+			continue
+		}
+
+		//separate key-val pairs in dctr string by pipe(|)
+		if dctr.String() != "" {
+			dctr.WriteString("|")
+		}
+
+		//trimming spaces from key
+		key = strings.TrimSpace(key)
+
+		switch typedValue := val.(type) {
+		case string:
+			fmt.Fprintf(&dctr, "%s=%s", key, strings.TrimSpace(typedValue))
+
+		case float64, bool:
+			fmt.Fprintf(&dctr, "%s=%v", key, typedValue)
+
+		case []interface{}:
+			if isStringArray(typedValue) {
+				if valStrArr := getStringArray(typedValue); valStrArr != nil && len(valStrArr) > 0 {
+					valStr := strings.Join(valStrArr[:], ",")
+					fmt.Fprintf(&dctr, "%s=%s", key, valStr)
+				}
+			}
+		}
+	}
+
+	if dctrStr := dctr.String(); dctrStr != "" {
+		extMap[dctrKeyName] = strings.TrimSuffix(dctrStr, "|")
+	}
+}
+
+//isStringArray check if []interface is a valid string array
+func isStringArray(array []interface{}) bool {
+	for _, val := range array {
+		if _, ok := val.(string); !ok {
+			return false
+		}
+	}
+	return true
+}
+
+//getStringArray converts interface of type string array to string array
+func getStringArray(val interface{}) []string {
+	aInterface, ok := val.([]interface{})
+	if !ok {
+		return nil
+	}
+	aString := make([]string, len(aInterface))
+	for i, v := range aInterface {
+		if str, ok := v.(string); ok {
+			aString[i] = strings.TrimSpace(str)
+		}
```

Comment:
To avoid needing the extra loop done by calling `isStringArray()`, you could have an else section of this if return nil if `v` is not of type `string`. It may be a minor performance enhancement.

It is also a little odd that `isStringArray()` takes an argument of `[]interface{}`, while `getStringArray()` takes `interface{}`, but both are passed the same value. The case statement above insures the value is of type `[]interface{}` which happens to also be a valid `interface{}`, so it works, but stuffs the `typedValue` into another interface abstraction which you then have to get out of.

Reply:
Removed `isStringArray` and reusing `getStringArray` for the same.
Changed type of `getStringArray` to `[]interface{}`